### PR TITLE
Adds after listener which does not run for nested scopes

### LINKED ISSUE #2792: Nested "before" lifecycle hook
I'm trying to use Kotest to write nested tests that have hooks that only run once per nest level.  This is how mocha works in JavaScript and how RSpec works in ruby.  In mocha the function is called `before`. I tried using `beforeContainer` but it runs once for *each* nested container instead of just once for the entire container including any nested containers within that parent container. This is hard to explain, so below is a set of tests that I would expect to pass, if a `before` method existed in kotest that behaved like mocha's or RSpec's.  In short, each `before` method should only be called once, and only when tests in its container run (e.g., the `before` method inside `"nested level 2"` shouldn't run until after the tests in `"nested level 1"` have run).

```kotlin
package foo.bar

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe

class WidgetTest : DescribeSpec({
    var foo = "initial"
    var counterOne = 0
    var counterTwo = 0
    var counterThree = 0

    before {
        foo = "bar"
        counterOne++
    }

    it("foo should be bar") {
        foo shouldBe "bar"
        counterOne shouldBe 1
    }

    it("beforeSpec should have been called only once") {
        foo shouldBe "bar"
        counterOne shouldBe 1
    }

    it("counterTwo should be 0") {
        counterTwo shouldBe 0
    }

    it("counterThree should be 0") {
        counterThree shouldBe 0
    }

    describe("nested level 1") {
        before {
            foo = "buzz"
            counterTwo++
        }

        it("foo should be buzz") {
            foo shouldBe "buzz"
        }

        it("and counterOne should be 1") {
            counterOne shouldBe 1
        }

        it("and counterTwo should be 1") {
            counterTwo shouldBe 1
        }

        describe("nested level 2") {
            before {
                foo = "jazz"
                counterThree++
            }

            it("foo should be jazz") {
                foo shouldBe "jazz"
            }

            it("and counterOne should be 1") {
                counterOne shouldBe 1
            }

            it("and counterTwo should be 1") {
                counterTwo shouldBe 1
            }

            it("and counterThree should be 1") {
                counterThree shouldBe 1
            }
        }
    }
})
```

I can get close to what I want if I make the first `before` a `beforeSpec` and the other two `beforeContainer` (and move them above each nested container), but then `counterTwo` gets incremented twice: once for `"nested level 1"` and once for `"nested level 2"`.  I'm not sure if I'm just doing something wrong or if this capability simply doesn't exist in kotest.

A somewhat more realistic example:

```kotlin
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import kotlinx.coroutines.delay

class WidgetApiTest : DescribeSpec({
    describe("CRUD operations") {
        var widget: Widget? = Widget(0, "")

        describe("when we create a widget") {
            before { // This should run exactly once
                createWidget(1, "widget-a")
            }

            describe("when we fetch the widget") {
                before { // This should run exactly once
                    widget = getWidget(1) // assume this is a slow HTTP or database request
                }

                it("the widget should not be null") {
                    widget shouldNotBe null
                }

                it("and the widget should have a name") {
                    widget!!.name shouldBe "widget-a"
                }

                describe("when we update the widget's name") {
                    before { // the above "before" functions should not have run again in this test container
                        widget?.let { w ->
                            w.name = "new-name"
                            updateWidget(w)
                        }
                    }

                    it("then the re-fetched widget should be updated") {
                        widget = getWidget(1) // assume this is a slow HTTP or database request
                        widget!!.name shouldBe "new-name"
                    }
                }
            }
        }
    }
})

data class Widget(val id: Int, var name: String?)

val widgetContainer: HashMap<Int, Widget> = hashMapOf()

suspend fun createWidget(id: Int, name: String?): Widget {
    delay(200) // slow HTTP request or Database call
    val widget = Widget(id, name)
    widgetContainer[id] = widget
    return widget
}

suspend fun getWidget(id: Int): Widget? {
    delay(200) // slow HTTP request or Database call
    return widgetContainer[id]
}

suspend fun updateWidget(widget: Widget): Widget {
    delay(200) // slow HTTP request or Database call
    val id = widget.id
    widgetContainer[id] = widget
    return widget
}
```

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #2792 #3979

### CODE REVIEW

File: kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt [hunk 1]
```diff
@@ -98,6 +98,23 @@ interface ContainerScope : TestScope {
       afterAny(f)
    }
 
+   /**
+    * Registers a callback that is executed once for this container after all tests have completed.
+    *
+    * This differs from [afterContainer] which is executed after each container including child containers.
+    */
+   fun after(f: suspend (TestCase) -> Unit) {
```
Comment (AlexCue987) [inline kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt]: should we rename `after` to `afterAllTestsInContainerCompleted`? I'm sure there is a better name than what I've suggested, but IMO just `after` is not clear enough

Comment (sksamuel) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt]: The name was the longest part of the discussion in #2792 
I'm open to anything, there's `afterScope` perhaps. I don't think we want something too wordy, like `afterContaineratThisLevelOnly`

Comment (AlexCue987) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt]: `afterScope` sounds good

Comment (OliverO2) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt]: I'm always interested in improving naming. In this case, I don't feel like I'm able to meaningfully contribute unless I have a complete visualization of our test hierarchy along with corresponding lifecycle methods (listeners, interceptors, whatever). Would be cool if we could get that documented completely in one place.

Comment (sksamuel) [reply kotest-framework/kotest-framework-api/src/commonMain/kotlin/io/kotest/core/spec/style/scopes/ContainerScope.kt]: afterScope it is
And will make a PR to  document all our callbacks.

**Review by AlexCue987 — APPROVED** [review]

**Comment by dondod** [comment]
I'm confused. Why was `afterScope` added but not `beforeScope`?  I know you can just put the code directly inside the block, but it's quite odd.

**Comment by sksamuel** [comment]
> I'm confused. Why was `afterScope` added but not `beforeScope`? I know you can just put the code directly inside the block, but it's quite odd.

It's literally because it wouldn't do anything that just putting the code at the same level does.
Do you think there's value to have a beforeScope block that is the same? If so, let me know why, we can consider it.

**Comment by dondod** [comment]
Thanks. Yeah, I think there's great value in adding it.  It provides a consistency that people might expect (consistency with other frameworks they might be coming from), and provides consistency within the framework (other "after" hooks have a corresponding "before" hook).  It also helps organize the code and make it obvious what is happening.  If I see code inside a `beforeScope` method, I know that it runs before the scope, and the corresponding `afterScope` runs after the scope.   I've had people at work ask me why I sometimes put code directly inside the `describe` block instead of in a `beforeEach` hook.  I explain that putting the code inside the `describe` block works like a "before all" hook.  They always ask, "why isn't there a `beforeAll` hook?".   Not having to explain that would be worth it to me :).  The `afterScope` hook is a huge win though.

**Comment by sksamuel** [comment]
Coming back to this and thinking about it - one of the issues with adding `beforeScope`, would be that people might assume it always runs before nested tests regardless of where it is defined, but it would be very complicated (and break existing behavior).

Eg,

```kotlin
context("foo") {
  test("bar") { }
  beforeScope { println("hello") }
}
```

Here it is reasonable to think that the `beforeScope` will run before bar, but it wouldn't as bar is executed as soon as it is discovered. So it would end up being counter intuitive I think. Whereas this looks like how it works.

```kotlin
context("foo") {
  test("bar") { }
  println("hello")
}
```

**Comment by dondod** [comment]
That's a good point, but I think it would be even more counterintuitive to put a `beforeScope` after a test at all.  In other test frameworks with a similar hook, I've never seen anyone put the "before" hook after a test, even if it would work (it would in rspec or mocha, for example).

On a side note, I recently had Claude Code refactor some `DescribeSpec` tests, and I told it to use `afterScope`, etc. to consolidate setup and cleanup.  It assumed there was a `beforeScope` and kept asking to search the kotest docs :).
